executor/hash_table.go

```diff
@@ -258,101 +259,137 @@ func (c *hashRowContainer) ActionSpill() memory.ActionOnExceed {
 
 const (
 	initialEntrySliceLen = 64
-	maxEntrySliceLen     = 8 * 1024
+	maxEntrySliceLen     = 8192
 )
 
 type entry struct {
 	ptr  chunk.RowPtr
-	next entryAddr
+	next *entry
 }
 
 type entryStore struct {
 	slices [][]entry
+	cursor int
 }
 
-func (es *entryStore) init() {
-	es.slices = [][]entry{make([]entry, 0, initialEntrySliceLen)}
-	// Reserve the first empty entry, so entryAddr{} can represent nullEntryAddr.
-	reserved := es.put(entry{})
-	if reserved != nullEntryAddr {
-		panic("entryStore: first entry is not nullEntryAddr")
-	}
+func newEntryStore() *entryStore {
+	es := new(entryStore)
+	es.slices = [][]entry{make([]entry, initialEntrySliceLen)}
+	es.cursor = 0
+	return es
 }
 
-func (es *entryStore) put(e entry) entryAddr {
+func (es *entryStore) GetStore() (e *entry) {
 	sliceIdx := uint32(len(es.slices) - 1)
 	slice := es.slices[sliceIdx]
-	if len(slice) == cap(slice) {
+	if es.cursor >= cap(slice) {
 		size := cap(slice) * 2
 		if size >= maxEntrySliceLen {
 			size = maxEntrySliceLen
 		}
-		slice = make([]entry, 0, size)
+		slice = make([]entry, size)
 		es.slices = append(es.slices, slice)
 		sliceIdx++
+		es.cursor = 0
 	}
-	addr := entryAddr{sliceIdx: sliceIdx, offset: uint32(len(slice))}
-	es.slices[sliceIdx] = append(slice, e)
-	return addr
-}
-
-func (es *entryStore) get(addr entryAddr) entry {
-	return es.slices[addr.sliceIdx][addr.offset]
+	e = &es.slices[sliceIdx][es.cursor]
+	es.cursor++
+	return
 }
 
-type entryAddr struct {
-	sliceIdx uint32
-	offset   uint32
+type baseHashTable interface {
+	Put(hashKey uint64, rowPtr chunk.RowPtr)
+	Get(hashKey uint64) (rowPtrs []chunk.RowPtr)
+	Len() uint64
 }
 
-var nullEntryAddr = entryAddr{}
-
-// rowHashMap stores multiple rowPtr of rows for a given key with minimum GC overhead.
+// unsafeHashTable stores multiple rowPtr of rows for a given key with minimum GC overhead.
 // A given key can store multiple values.
 // It is not thread-safe, should only be used in one goroutine.
-type rowHashMap struct {
-	entryStore entryStore
-	hashTable  map[uint64]entryAddr
-	length     int
+type unsafeHashTable struct {
+	hashMap    map[uint64]*entry
+	entryStore *entryStore
+	length     uint64
+	keepOrder  bool
```

Comment:
The field is necessary? The old `rowHashMap` seems keepOrder always?

Reply:
It keeps order for the old `rowHashMap`, i.e., `unsafeHashTable`.
the `concurrentMap` cannot keep order due to concurrent inserting.